p2p/protocols/protocol.go

```diff
@@ -208,17 +201,27 @@ type Peer struct {
 	*p2p.Peer                   // the p2p.Peer object representing the remote
 	rw        p2p.MsgReadWriter // p2p.MsgReadWriter to send messages to and read messages from
 	spec      *Spec
+	encode    func(context.Context, interface{}) (interface{}, int, error)
+	decode    func(p2p.Msg) (context.Context, []byte, error)
 }
 
 // NewPeer constructs a new peer
 // this constructor is called by the p2p.Protocol#Run function
 // the first two arguments are the arguments passed to p2p.Protocol.Run function
 // the third argument is the Spec describing the protocol
-func NewPeer(p *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
+func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
+	encode := encodeWithContext
+	decode := decodeWithContext
+	if spec.DisableContext || !tracing.Enabled {
```

Comment:
Argument `spec` can be nil, and it should be handled in a way that DisableContext is considered as default value `false`.

Reply:
good catch thanks, but I think spec can only be nil in tests, where the default should be true

Reply:
In that case, maybe it should be renamed to EnableContext (with default false), as @skylenet suggested. This way, nil or &Spec{} (any instace without specifying the field) would have the same meaning.

Reply:
But, in this case, just handling nil spec is probably simpler. A reverse flag would make sense for structure with more fields.

Reply:
in fact a nil spec if just a testing feature, maybe shouldnt be allowed really.
anyhow i wanted to make the most minimal change.
up till this PR the spec didnt specify this field, so the default bool value should be  the default for production which is to have tracing enabled. 
If i change  it all ptotocoll specs will need to include `EnableContext` set to true.
Do you prefer this?

Reply:
No, current state is fine.